subsys/net/lib/tls_conf/tls_conf.c

```diff
@@ -0,0 +1,290 @@
+ * insecure).
+ */
+static int ztls_entropy_func(void *data, unsigned char *output, size_t len)
+{
+	size_t i = len / 4;
+	u32_t v;
+
+	while (i--) {
+		v = sys_rand32_get();
+		UNALIGNED_PUT(v, (u32_t *)output);
+		output += 4;
+	}
+
+	i = len & 3;
+	v = sys_rand32_get();
+	while (i--) {
+		*output++ = v;
+		v >>= 8;
+	}
+
+	return 0;
+}
+#endif
+
+static int ztls_mbedtls_ctr_drbg_random(void *p_rng, unsigned char *output,
+					size_t output_len)
+{
+	static K_MUTEX_DEFINE(mutex);
+	int res;
+
+	/* Avoid connection lockups due to no entropy, error out instead */
+	res = k_mutex_lock(&mutex, K_SECONDS(1));
+	if (res < 0) {
+		return res;
+	}
+
+	res = mbedtls_ctr_drbg_random(p_rng, output, output_len);
+	k_mutex_unlock(&mutex);
+
+	return res;
+}
+
+#else /* __ZEPHYR__ */
+
+static int null_entropy_func(void *data, unsigned char *output, size_t len)
+{
+	/* Warning: this is test-only implementation for non-Zephyr
+	 * environments.
+	 */
+	(void)data;
+	(void)output;
+	(void)len;
+	return 0;
+}
+
+#define ztls_mbedtls_ctr_drbg_random mbedtls_ctr_drbg_random
+
+#endif
+
+#ifdef MBEDTLS_DEBUG_C
+static void ztls_mbedtls_debug(void *ctx, int level, const char *file,
+			       int line, const char *str)
+{
+	(void)ctx;
+	printk("MBEDTLS%d:%s:%04d: %s\n", level, file, line, str);
+}
+#endif
+
+#define ztls_system_is_inited() (ztls_ctr_drbg.f_entropy != NULL)
+
+
+static int ztls_system_init(void)
+{
+	int ret;
+	/* TODO: Should use something device-specific, like MAC address */
+	static const unsigned char drbg_seed[] = "zephyr";
+	struct device *dev = NULL;
+
+#if defined(__ZEPHYR__) && defined(CONFIG_ENTROPY_HAS_DRIVER)
+	dev = device_get_binding(CONFIG_ENTROPY_NAME);
+
+	if (!dev) {
+		SYS_LOG_ERR("can't get entropy device");
+		return -ENODEV;
+	}
+#else
+	printk("*** WARNING: This system lacks entropy driver, "
+	       "TLS communication may be INSECURE! ***\n\n");
+#endif
+
+	/* We don't use mbedTLS entropy pool as of now. */
+	/* mbedtls_entropy_init(&p->entropy); */
+
+	mbedtls_ctr_drbg_init(&ztls_ctr_drbg);
+#ifdef __ZEPHYR__
+	ret = mbedtls_ctr_drbg_seed(&ztls_ctr_drbg, ztls_entropy_func, dev,
+				    drbg_seed, sizeof(drbg_seed));
+#else
+	ret = mbedtls_ctr_drbg_seed(&ztls_ctr_drbg, null_entropy_func, NULL,
+				    drbg_seed, sizeof(drbg_seed));
+#endif
+	if (ret != 0) {
+		mbedtls_ctr_drbg_free(&ztls_ctr_drbg);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ztls_init_tls_conf(mbedtls_ssl_config *conf, int client_or_serv)
+{
+	int ret;
+
+	if (!ztls_system_is_inited()) {
+		ret = ztls_system_init();
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	mbedtls_ssl_config_init(conf);
+
+	ret = mbedtls_ssl_config_defaults(conf, client_or_serv,
+					  MBEDTLS_SSL_TRANSPORT_STREAM,
+					  MBEDTLS_SSL_PRESET_DEFAULT);
+	if (ret != 0) {
+		SYS_LOG_ERR("mbedtls_ssl_config_defaults returned -0x%x",
+			    -ret);
+		goto cleanup_conf;
+	}
+
+	mbedtls_ssl_conf_rng(conf, ztls_mbedtls_ctr_drbg_random,
+			     &ztls_ctr_drbg);
+
+	#if 0
```

Comment:
?

Reply:
This is kind of multi-line code comment, next line says:

~~~
/* These are defaults in mbedTLS */
~~~

Given that this stuff is related to security, I think it's good idea to have a comment like that, essentially saying "we don't include code like that to save on code space, because those are mbedTLS defaults, but these defaults exactly are: [setting cert verification as required for client connections], [setting client cert verification as not done for server connections]".

If we'd allow `//` comments, I'd use them instead of `#if 0`, and using `/* */` comments doesn't work too well for code, because in our code style they're more intended for pure textual comments.

If you have any suggestions, I'd appreciate them. (For any other code I'd just remove that block, but per above, that's security, and IMHo it should be as explicit and detailed as possible.)